### PR TITLE
First attempt, not tested yet

### PR DESCRIPTION
**Pull request by einaregilsson**
Exponential backoff in the aws-api-request.js itself, so applies to all operations.

### CODE REVIEW

**Comment by einaregilsson** [comment]
Still very much not working, but I think this is the right approach. Need to test locally.

File: aws-api-request.js [hunk 1]
```diff
@@ -141,14 +174,21 @@ function createResult(data, res) {
...
     delete headers.Host;
     headers['Content-Length'] = data.length;
     const port = 443;
+
+    //Test of exponential backoff, use at least 3 attempts for every operation ...
+    if (retryAttempt < 3) {
+    //    callback(null, { statusCode: 400, data : { Error : { Code : 'Throttling'}}});
+    //    return;
+    }
+
```
Comment (n-hebert) [inline aws-api-request.js:191]: At first look, this is redundant and we can remove it -- the above should take care of it all.

Is that correct?

File: aws-api-request.js [hunk 2]
```diff
@@ -122,6 +124,33 @@ function awsApiRequest(options) {
                         ...options,
                         host: url.hostname
                     }));
+                } else if (wasThrottled(result)) {
+                    //Exponential backoff with a 500ms jitter
```
Comment (n-hebert) [inline aws-api-request.js:128]: Jitter! Very neat.

**Review by n-hebert — COMMENTED** [review]
This does look a lot cleaner.

I'll give it some tries locally and online!

**Comment by n-hebert** [comment]
This is unused now, btw 
https://github.com/einaregilsson/beanstalk-deploy/blob/0df1c963a04f4b20251ec66c09e29022a95ee2da/beanstalk-deploy.js#L375

**Review by n-hebert — APPROVED** [review]
[It's working!](https://media.giphy.com/media/CuMiNoTRz2bYc/giphy.gif)

```
16:39:40 INFO: Environment update is starting.
16:39:47 INFO: Deploying new version to instance(s).
Request for DescribeEnvironments in elasticbeanstalk was throttled (retry attempt 0). Retrying in 462ms...
Request for DescribeEvents in elasticbeanstalk was throttled (retry attempt 0). Retrying in 140ms...
Request for DescribeEvents in elasticbeanstalk was throttled (retry attempt 1). Retrying in 551ms...
Request for DescribeEnvironments in elasticbeanstalk was throttled (retry attempt 1). Retrying in 333ms...
Request for DescribeEvents in elasticbeanstalk was throttled (retry attempt 2). Retrying in 744ms...
Request for DescribeEnvironments in elasticbeanstalk was throttled (retry attempt 2). Retrying in 554ms...
Request for DescribeEnvironments in elasticbeanstalk was throttled (retry attempt 3). Retrying in 1232ms...
Request for DescribeEvents in elasticbeanstalk was throttled (retry attempt 3). Retrying in 1294ms...
Request for DescribeEnvironments in elasticbeanstalk was throttled (retry attempt 4). Retrying in 1835ms...
Request for DescribeEvents in elasticbeanstalk was throttled (retry attempt 4). Retrying in 2067ms...
Request for DescribeEvents in elasticbeanstalk was throttled (retry attempt 5). Retrying in 3665ms...
```

**Comment by n-hebert** [comment]
I've made a couple tidies of the formatting along the way and clipped the unused function, if you'd like to check out my latest commit on the equivalent branch over here: https://github.com/rewindio/beanstalk-deploy/pull/3/commits

**Comment by einaregilsson** [comment]
@n-hebert Thanks, I pulled the commit from your fork. Doing some final testing, will publish later today.

**Pull request merged**